File: sdk/wasm/operation_log.go
//go:build js && wasm
// +build js,wasm

package main

import (
	"encoding/json"
	"strings"
	"syscall/js"

	"github.com/sat20-labs/sat20wallet/sdk/wallet"
)

const operationLogModule = "sat20wallet_operation_log"

type operationLogCreateRequest struct {
	Category   string            `json:"category"`
	Action     string            `json:"action"`
	Title      string            `json:"title"`
	Summary    string            `json:"summary"`
	Parameters map[string]string `json:"parameters"`
}

type operationLogUpdateRequest struct {
	Status  string            `json:"status"`
	Message string            `json:"message"`
	Details map[string]string `json:"details"`
	Result  map[string]string `json:"result"`
	TxID    string            `json:"txid"`
}

func beginOperationLog(this js.Value, p []js.Value) any {
	if _mgr == nil {
		return createJsRet(nil, -1, "Manager not initialized")
	}
	if len(p) != 1 || p[0].Type() != js.TypeString || strings.TrimSpace(p[0].String()) == "" {
		return createJsRet(nil, -1, "operation log request is required")
	}
	var req operationLogCreateRequest
	if err := json.Unmarshal([]byte(p[0].String()), &req); err != nil {
		return createJsRet(nil, -1, err.Error())
	}
	handler := createAsyncJsHandler(func() (interface{}, int, string) {
		record, err := _mgr.BeginOperationLog(wallet.OperationLogCreate{
			Category:   req.Category,
			Action:     req.Action,
			Title:      req.Title,
			Summary:    req.Summary,
			Parameters: req.Parameters,
		})
		if err != nil {
			return nil, -1, err.Error()
		}
		return map[string]any{"id": record.ID}, 0, "ok"
	})
	return js.Global().Get("Promise").New(handler)
}

func updateOperationLog(this js.Value, p []js.Value) any {
	if _mgr == nil {
		return createJsRet(nil, -1, "Manager not initialized")
	}
	if len(p) != 2 || p[0].Type() != js.TypeString || strings.TrimSpace(p[0].String()) == "" ||
		p[1].Type() != js.TypeString || strings.TrimSpace(p[1].String()) == "" {
		return createJsRet(nil, -1, "operation log id and update are required")
	}
	id := strings.TrimSpace(p[0].String())
	var req operationLogUpdateRequest
	if err := json.Unmarshal([]byte(p[1].String()), &req); err != nil {
		return createJsRet(nil, -1, err.Error())
	}
	handler := createAsyncJsHandler(func() (interface{}, int, string) {
		record, err := _mgr.UpdateOperationLog(id, wallet.OperationLogUpdate{
			Status:  wallet.OperationLogStatus(req.Status),
			Message: req.Message,
			Details: req.Details,
			Result:  req.Result,
			TxID:    req.TxID,
		})
		if err != nil {
			return nil, -1, err.Error()
		}
		encoded, err := json.Marshal(record)
		if err != nil {
			return nil, -1, err.Error()
		}
		return map[string]any{"log": string(encoded)}, 0, "ok"
	})
	return js.Global().Get("Promise").New(handler)
}

func getOperationLogs(this js.Value, p []js.Value) any {
	if _mgr == nil {
		return createJsRet(nil, -1, "Manager not initialized")
	}
	handler := createAsyncJsHandler(func() (interface{}, int, string) {
		items, err := _mgr.GetOperationLogs()
		if err != nil {
			return nil, -1, err.Error()
		}
		encoded, err := json.Marshal(items)
		if err != nil {
			return nil, -1, err.Error()
		}
		return map[string]any{"logs": string(encoded)}, 0, "ok"
	})
	return js.Global().Get("Promise").New(handler)
}

func getOperationLog(this js.Value, p []js.Value) any {
	if _mgr == nil {
		return createJsRet(nil, -1, "Manager not initialized")
	}
	if len(p) != 1 || p[0].Type() != js.TypeString || strings.TrimSpace(p[0].String()) == "" {
		return createJsRet(nil, -1, "operation log id is required")
	}
	id := strings.TrimSpace(p[0].String())
	handler := createAsyncJsHandler(func() (interface{}, int, string) {
		item, err := _mgr.GetOperationLog(id)
		if err != nil {
			return nil, -1, err.Error()
		}
		if item == nil {
			return nil, -1, "operation log not found"
		}
		encoded, err := json.Marshal(item)
		if err != nil {
			return nil, -1, err.Error()
		}
		return map[string]any{"log": string(encoded)}, 0, "ok"
	})
	return js.Global().Get("Promise").New(handler)
}

func deleteAllOperationLogs(this js.Value, p []js.Value) any {
	if _mgr == nil {
		return createJsRet(nil, -1, "Manager not initialized")
	}
	handler := createAsyncJsHandler(func() (interface{}, int, string) {
		if err := _mgr.DeleteAllOperationLogs(); err != nil {
			return nil, -1, err.Error()
		}
		return nil, 0, "ok"
	})
	return js.Global().Get("Promise").New(handler)
}

func init() {
	obj := js.Global().Get("Object").New()
	obj.Set("beginOperationLog", js.FuncOf(beginOperationLog))
	obj.Set("updateOperationLog", js.FuncOf(updateOperationLog))
	obj.Set("getOperationLogs", js.FuncOf(getOperationLogs))
	obj.Set("getOperationLog", js.FuncOf(getOperationLog))
	obj.Set("deleteAllOperationLogs", js.FuncOf(deleteAllOperationLogs))
	js.Global().Set(operationLogModule, obj)
}

File: pwa/utils/operationLog.ts
export type OperationLogStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled'

export interface OperationLogEvent {
  timestamp: number
  status?: OperationLogStatus
  message: string
  details?: Record<string, string>
}

export interface OperationLogRecord {
  id: string
  category: string
  action: string
  title: string
  summary: string
  status: OperationLogStatus
  created_at: number
  updated_at: number
  reservation_id?: number
  reservation_type?: string
  txid?: string
  parameters?: Record<string, string>
  result?: Record<string, string>
  history: OperationLogEvent[]
}

export interface OperationLogCreateInput {
  category: string
  action: string
  title: string
  summary: string
  parameters?: Record<string, string>
}

export interface OperationLogUpdateInput {
  status?: OperationLogStatus
  message: string
  details?: Record<string, string>
  result?: Record<string, string>
  txid?: string
}

type WasmResult = {
  code: number
  msg?: string
  data?: Record<string, unknown>
}

function moduleApi(): any {
  return (globalThis as any).sat20wallet_operation_log
}

async function invoke(method: string, ...args: unknown[]): Promise<[Error | null, Record<string, unknown> | null]> {
  const api = moduleApi()
  if (!api || typeof api[method] !== 'function') {
    return [new Error('Operation log module is unavailable'), null]
  }
  try {
    const result = await api[method](...args) as WasmResult
    if (!result || result.code !== 0) {
      return [new Error(result?.msg || `${method} failed`), null]
    }
    return [null, result.data || null]
  } catch (error) {
    return [error instanceof Error ? error : new Error(String(error)), null]
  }
}

export async function beginOperationLog(input: OperationLogCreateInput): Promise<[Error | null, string | null]> {
  const [error, data] = await invoke('beginOperationLog', JSON.stringify(input))
  if (error) return [error, null]
  const id = typeof data?.id === 'string' ? data.id : ''
  return id ? [null, id] : [new Error('Operation log id is missing'), null]
}

export async function updateOperationLog(id: string, input: OperationLogUpdateInput): Promise<Error | null> {
  if (!id) return null
  const [error] = await invoke('updateOperationLog', id, JSON.stringify(input))
  return error
}

export async function getOperationLogs(): Promise<[Error | null, OperationLogRecord[]]> {
  const [error, data] = await invoke('getOperationLogs')
  if (error) return [error, []]
  try {
    const raw = typeof data?.logs === 'string' ? data.logs : '[]'
    return [null, JSON.parse(raw) as OperationLogRecord[]]
  } catch (parseError) {
    return [parseError instanceof Error ? parseError : new Error(String(parseError)), []]
  }
}

export async function getOperationLog(id: string): Promise<[Error | null, OperationLogRecord | null]> {
  const [error, data] = await invoke('getOperationLog', id)
  if (error) return [error, null]
  try {
    const raw = typeof data?.log === 'string' ? data.log : ''
    if (!raw) return [new Error('Operation log not found'), null]
    return [null, JSON.parse(raw) as OperationLogRecord]
  } catch (parseError) {
    return [parseError instanceof Error ? parseError : new Error(String(parseError)), null]
  }
}

export async function deleteAllOperationLogs(): Promise<Error | null> {
  const [error] = await invoke('deleteAllOperationLogs')
  return error
}
